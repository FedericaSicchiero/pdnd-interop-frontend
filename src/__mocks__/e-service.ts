import { EServiceReadType } from '../../types'

export const eserviceDraft: EServiceReadType = {
  producer: {
    id: 'djofsi-sdfjdsi-djsfs',
    name: 'Comune di Bologna',
  },
  name: 'Nome e-service',
  description: 'Descrizione e-service',
  technology: 'REST',
  id: 'sdjof-sdfjdspof-dsfdsjf',
  state: 'DRAFT',
  descriptors: [],
  attributes: {
    verified: [],
    declared: [],
    certified: [
      {
        single: {
          id: 'dsdsld-dsdlds-lsdasdas',
          creationTime: '2022-02-28T16:16:18.879093Z',
          explicitAttributeVerification: false,
          verified: false,
          origin: 'dfkdsfk',
          code: 'dfjdso',
          name: 'Attributo 1',
          description: 'Descrizione attributo 1',
        },
      },
    ],
  },
}
