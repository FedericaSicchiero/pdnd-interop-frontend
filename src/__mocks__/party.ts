import { Party } from '../../types'

export const partyActiveManager: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'MANAGER',
  productInfo: { createdAt: '', id: 'interop', role: 'admin' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}

export const partyActiveDelegate: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'DELEGATE',
  productInfo: { createdAt: '', id: 'interop', role: 'admin' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}

export const partyActiveOperatorApi: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'OPERATOR',
  productInfo: { createdAt: '', id: 'interop', role: 'api' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}

export const partyActiveOperatorSecurity: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'OPERATOR',
  productInfo: { createdAt: '', id: 'interop', role: 'security' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}
