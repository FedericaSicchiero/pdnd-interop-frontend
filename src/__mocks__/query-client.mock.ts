import { QueryClient, QueryClientConfig } from '@tanstack/react-query'
import { queryClientConfig } from '../config/query-client'
import { deepmerge } from '@mui/utils'
import noop from 'lodash/noop'

const queryClientConfigMock: QueryClientConfig = deepmerge(
  {
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
    // Disables logging
    logger: {
      log: noop,
      warn: noop,
      error: noop,
    },
  },
  queryClientConfig
)

export const queryClientMock = new QueryClient(queryClientConfigMock)
