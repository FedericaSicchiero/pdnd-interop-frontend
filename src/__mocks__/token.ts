export const token = process.env.REACT_APP_MOCK_TOKEN as string
