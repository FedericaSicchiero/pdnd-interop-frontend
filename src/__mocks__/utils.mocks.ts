import { Paginated } from './../api/comon.api.types'

/** Wraps the data passed in the pagination data structure */
export function createPaginatedMockData<T>(
  data: Array<T>,
  pagination?: Partial<Paginated<T>['pagination']>
): Paginated<T> {
  return {
    results: data,
    pagination: {
      offset: pagination?.offset ?? 0,
      limit: pagination?.offset ?? 50,
      totalResults: pagination?.offset ?? data.length,
    },
  }
}
