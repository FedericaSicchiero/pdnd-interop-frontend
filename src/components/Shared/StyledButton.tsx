import React, { FunctionComponent } from 'react'
import { Button } from '@mui/material'

export const StyledButton: FunctionComponent<any> = React.forwardRef(
  ({ children, ...props }, ref) => {
    return (
      <Button {...props} ref={ref}>
        {children}
      </Button>
    )
  }
)
