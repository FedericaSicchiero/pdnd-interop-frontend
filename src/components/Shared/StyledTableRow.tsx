import React, { FunctionComponent, ReactElement } from 'react'
import { TableCell, TableRow, Typography } from '@mui/material'

type LabelCell = {
  label: string | JSX.Element
  tooltip?: ReactElement
}

type CustomCell = {
  custom: React.ReactNode
}

type Cell = LabelCell | CustomCell

type StyledTableRowProps = {
  cellData: Array<Cell>
}

export const StyledTableRow: FunctionComponent<StyledTableRowProps> = ({ cellData, children }) => {
  return (
    <TableRow sx={{ bgcolor: 'common.white' }}>
      {cellData.map((cell, i) => {
        return (
          <TableCell key={i} sx={{ py: 2 }}>
            {'custom' in cell ? (
              cell.custom
            ) : (
              <Typography component="span" sx={{ display: 'inline-block' }} variant="body2">
                {cell.label}
                {cell.tooltip}
              </Typography>
            )}
          </TableCell>
        )
      })}
      {children && (
        <TableCell sx={{ minWidth: '15rem' }} align="right">
          {children}
        </TableCell>
      )}
    </TableRow>
  )
}
