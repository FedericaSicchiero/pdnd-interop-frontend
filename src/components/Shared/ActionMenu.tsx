import React, { FunctionComponent, useContext, useRef, useState } from 'react'
import { MoreVert as MoreVertIcon } from '@mui/icons-material'
import { StyledButton } from './StyledButton'
import { Menu, MenuItem } from '@mui/material'
import { ActionProps, ExtendedMUIColor } from '../../../types'
import { TableActionMenuContext } from '../../lib/context'
import { Box } from '@mui/system'

type ActionMenuProps = {
  // The list of actions to display in the menu
  actions: Array<ActionProps>
  onOpen: (id: string) => void
  onClose: VoidFunction
  // The id of the menu currently clicked. In a table there may be many of them,
  // but only one is open at a time (or none, if the value is null)
  openMenuId: string | null
  // Only used for snapshot tests, to have a stable id
  snapshotTestInternalId?: string
  iconColor?: ExtendedMUIColor
}

export const ActionMenu: FunctionComponent<
  Omit<ActionMenuProps, 'onOpen' | 'onClose' | 'openMenuId'>
> = (props) => {
  const { tableActionMenu, setTableActionMenu } = useContext(TableActionMenuContext)

  const onOpen = (id: string) => {
    setTableActionMenu(id)
  }

  const onClose = () => {
    setTableActionMenu(null)
  }

  return (
    <ActionMenuComponent
      {...props}
      onOpen={onOpen}
      onClose={onClose}
      openMenuId={tableActionMenu}
    />
  )
}

// From https://stackoverflow.com/a/1349426
function uniqueString(length: number) {
  let result = ''
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  for (let i = 0; i < length; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length))
  }
  return result
}

const ActionMenuComponent: FunctionComponent<ActionMenuProps> = ({
  actions,
  onOpen,
  onClose,
  openMenuId,
  snapshotTestInternalId,
  iconColor = 'primary',
}) => {
  // Needs to be state to avoid it changing on rerender
  const [id] = useState(snapshotTestInternalId || uniqueString(8))
  const anchorRef = useRef() as React.MutableRefObject<HTMLSpanElement>
  const anchorId = `basic-button-${id}`
  const menuId = `basic-menu-${id}`
  const open = Boolean(openMenuId !== null && openMenuId === anchorId)

  const handleClick = (event: React.SyntheticEvent) => {
    event.stopPropagation()
    event.preventDefault()
    onOpen(event.currentTarget.id)
  }

  if (!Boolean(actions.length > 0)) {
    // Used to keep the buttons visually aligned in case there is no ActionMenu
    return <Box component="span" sx={{ width: 70, display: 'inline-block' }} />
  }

  return (
    <React.Fragment>
      <StyledButton
        id={anchorId}
        aria-controls={menuId}
        aria-haspopup="true"
        aria-expanded={open ? 'true' : undefined}
        onClick={handleClick}
        color={iconColor}
      >
        <span ref={anchorRef}>
          <MoreVertIcon color="inherit" fontSize="small" />
        </span>
      </StyledButton>

      <Menu
        id={menuId}
        anchorEl={anchorRef.current}
        open={open}
        onClose={onClose}
        MenuListProps={{
          'aria-labelledby': 'basic-button',
        }}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      >
        {actions.map(({ onClick, label }, i) => {
          return (
            <MenuItem key={i} onClick={onClick}>
              {label}
            </MenuItem>
          )
        })}
      </Menu>
    </React.Fragment>
  )
}
