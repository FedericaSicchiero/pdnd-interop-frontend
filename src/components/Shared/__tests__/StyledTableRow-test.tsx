import React from 'react'
import renderer from 'react-test-renderer'
import { noop } from 'lodash'
import { createMemoryHistory } from 'history'
import { StyledTableRow } from '../StyledTableRow'
import { StyledButton } from '../StyledButton'
import { AllTheProviders } from '../../../__mocks__/providers'
import { ActionMenu } from '../ActionMenu'

const cellData = [{ label: 'Mario' }, { label: 'Rossi' }, { label: 'rejsedf3-re4k-rew2-eoer' }]
const singleActionLink = { label: 'Ispeziona', to: '/rotta-esempio' }
const actions = [
  { onClick: noop, label: 'Azione 1' },
  { onClick: noop, label: 'Azione 2' },
]

describe('Snapshot', () => {
  it('matches row without actions', () => {
    const component = renderer.create(<StyledTableRow cellData={cellData} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('matches row with single action button', () => {
    const history = createMemoryHistory()
    const component = renderer.create(
      <AllTheProviders defaultHistory={history}>
        <StyledTableRow cellData={cellData}>
          <StyledButton
            onClick={() => {
              history.push(singleActionLink.to)
            }}
          >
            {singleActionLink.label}
          </StyledButton>
        </StyledTableRow>
      </AllTheProviders>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('matches row with single action button and action list button', () => {
    const history = createMemoryHistory()
    const component = renderer.create(
      <AllTheProviders defaultHistory={history}>
        <StyledTableRow cellData={cellData}>
          <StyledButton
            onClick={() => {
              history.push(singleActionLink.to)
            }}
          >
            {singleActionLink.label}
          </StyledButton>
        </StyledTableRow>

        <ActionMenu actions={actions} />
      </AllTheProviders>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
