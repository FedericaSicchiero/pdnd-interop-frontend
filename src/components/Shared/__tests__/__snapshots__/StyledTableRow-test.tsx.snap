// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches row with single action button 1`] = `
<tr
  className="MuiTableRow-root css-1p72hwd-MuiTableRow-root"
  role={null}
>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
  >
    <span
      className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
    >
      Mario
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
  >
    <span
      className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
    >
      Rossi
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
  >
    <span
      className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
    >
      rejsedf3-re4k-rew2-eoer
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium css-1mrogcl-MuiTableCell-root"
  >
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Ispeziona
    </button>
  </td>
</tr>
`;

exports[`Snapshot matches row with single action button and action list button 1`] = `
Array [
  <tr
    className="MuiTableRow-root css-1p72hwd-MuiTableRow-root"
    role={null}
  >
    <td
      aria-sort={null}
      className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
      >
        Mario
      </span>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
      >
        Rossi
      </span>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
      >
        rejsedf3-re4k-rew2-eoer
      </span>
    </td>
    <td
      aria-sort={null}
      className="MuiTableCell-root MuiTableCell-alignRight MuiTableCell-sizeMedium css-1mrogcl-MuiTableCell-root"
    >
      <button
        className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Ispeziona
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </td>
  </tr>,
  <button
    aria-controls="basic-menu-1"
    aria-haspopup="true"
    className="MuiButton-root MuiButton-naked MuiButton-nakedPrimary MuiButton-sizeSmall MuiButton-nakedSizeSmall MuiButtonBase-root  css-19lhi6j-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    id="basic-button-1"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall css-1bptiuh-MuiSvgIcon-root"
        data-testid="MoreVertIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </span>
  </button>,
]
`;

exports[`Snapshot matches row without actions 1`] = `
<tr
  className="MuiTableRow-root css-1p72hwd-MuiTableRow-root"
  role={null}
>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
  >
    <span
      className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
    >
      Mario
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
  >
    <span
      className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
    >
      Rossi
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-sizeMedium css-hxqovd-MuiTableCell-root"
  >
    <span
      className="MuiTypography-root MuiTypography-body2 css-1h9j9ew-MuiTypography-root"
    >
      rejsedf3-re4k-rew2-eoer
    </span>
  </td>
</tr>
`;
